runtime/src/iree/vm/module.h

```diff
@@ -397,15 +397,29 @@ typedef struct iree_vm_module_t {
                                       iree_vm_signal_t signal);
 
   // Begins a function call with the given |call| arguments.
-  // Execution may yield in the case of asynchronous code and require one or
-  // more calls to the resume method to complete.
+  //
+  // Returns OK if execution completes immediately. If the call completes
+  // immediately the results will be written to |call|->results.
+  //
+  // Returns IREE_STATUS_DEFERRED if execution yielded and the call needs to be
+  // resumed. Depending on the program it may be unsafe to begin any other calls
```

Comment:
Can you say more (at least to me) about this scary looking admonition?

Reply:
"unsafe" here is more about correctness than security/etc. It's mostly about there being no way for the runtime to enforce that the program being run uses coroutines properly - it has to be built to do the right thing based on the intent. An example would be global state - if a coroutine loads globals, yields, and then stores globals it'd not be safe (in the correctness sense) to launch other coroutines that touch those same globals. IOW the runtime supports an arbitrary number of simultaneously active coroutines but the program is what defines those and how they work.

For programs that come through the stream dialect we'll make sure things work mostly right (when I get semaphores plumbed through, `Any Day Now(tm)`) - you'll be able to invoke a function any number of times and the invocations will order themselves correctly based on the semaphores when everything is working. But if you hand-crafted a program or lowered from some other path it may not handle things right unless they accounted for the same. There's still some work to finish up in the compiler for this but it'll all be independent from the runtime.

Still some sketching to do for how this surfaces on the `iree_vm_invocation_t` (the coroutine/call state API you'll be wrapping) - my hope is that we don't end up with the runtime needing to serialize things based on module metadata but that's always an option ("export @foo cannot run unless the queue is empty" vs "export @bar is safe to execute under any circumstance"). To start I'll likely serialize everything as that's easiest pending the TBD compiler work which still allows for pipelining but not for concurrent execution within the same context (ala goroutines). Independent contexts will always be able to run concurrently, though (no globals or state hazards to worry about) - so two instances of model A can interleave as can an instance each of model A and model B. We may also decide here to say that every invocation in a context is serialized and leave the goroutine stuff for internals accessible by the compiler instead of a hosting user.

I'm realizing this may not help - but once I get this API baked I'll do a bigger writeup with diagrams (or tracy captures) :)